ppafm/ocl/field.py

```diff
@@ -850,101 +1012,171 @@ def calc_force_fft(self, FE=None, rot=np.eye(3), rot_center=np.zeros(3), local_s
-            FF = self.run_evalLJC_Q( FE=FE, Qmix=Qmix, local_size=(32,), bCopy=bCopy, bFinish=bFinish )
+        # Interpolate sample Hartree potential and electron density onto the correct grid
+        lvec = np.concatenate([self.lvec0[None, :3], self.lvec[:, :3]], axis=0)
+        pot_lvec_same = (np.allclose(lvec, self.pot.lvec) and np.allclose(self.pot.shape, self.nDim[:3]))
+        rho_sample_lvec_same = (np.allclose(lvec, self.rho_sample.lvec) and np.allclose(self.rho_sample.shape, self.nDim[:3]))
+        if not pot_lvec_same:
+            pot = self.pot.interp_at(lvec, self.nDim[:3], rot=rot, rot_center=rot_center,
+                local_size=local_size, queue=self.queue)
         else:
-            if np.allclose(self.atoms[:, -1], 0): # No charges
-                FF = self.run_evalLJ_noPos()
-            elif bQZ:
-                FF = self.run_evalLJC_QZs_noPos( FE=FE, Qmix=Qmix, local_size=(32,), bCopy=bCopy, bFinish=bFinish )
-            else:
-                FF = self.run_evalLJC_Q_noPos( FE=FE, Qmix=Qmix, local_size=(32,), bCopy=bCopy, bFinish=bFinish )
+            pot = self.pot
+        if not rho_sample_lvec_same:
+            rho_sample = self.rho_sample.interp_at(lvec, self.nDim[:3], rot=rot, rot_center=rot_center,
+                local_size=local_size, queue=self.queue)
+        else:
+            rho_sample = self.rho_sample
 
-        if(bRelease): self.tryReleaseBuffers()
-        if(bRuntime): print("runtime(ForceField_LJC.makeFF.tot) [s]: ", time.time() - t0)
+        if bRuntime:
+            self.queue.finish()
+            print("runtime(ForceField_LJC.calc_force_fdbm.interpolate) [s]: ", time.perf_counter() - t0)
 
-        return FF, atoms
+        # Cross-correlate Hartree potential and tip electron delta density for electrostatic energy
+        E_es = self.fft_corr_delta.correlate(pot)
+        if not pot_lvec_same: pot.release()
 
-    def makeFFHartree(self, atoms, cLJs, pot=None, rho=None, FE=None, rot=np.eye(3), rot_center=np.zeros(3),
+        # Cross-correlate sample electron density and tip electron density for Pauli energy
+        if not np.allclose(B, 1.0):
+            rho_sample = rho_sample.power_positive(p=B, in_place=False)
+        E_pauli = self.fft_corr.correlate(rho_sample)
+        if not rho_sample_lvec_same: rho_sample.release()
+
+        if bRuntime:
+            self.queue.finish()
+            print("runtime(ForceField_LJC.calc_force_fdbm.cross-correlation) [s]: ", time.perf_counter() - t0)
+
+        # Add the two energy contributions together
+        E = E_es.add_mult(E_pauli, scale=A, in_place=True, local_size=local_size, queue=self.queue)
+        E_pauli.release()
+
+        if bRuntime:
+            self.queue.finish()
+            print("runtime(ForceField_LJC.calc_force_fdbm.add_mult) [s]: ", time.perf_counter() - t0)
+
+        # Take gradient to get the force field
+        E.grad(scale=[-1.0, -1.0, -1.0, 1.0], array_out=self.cl_FE, order='F',
```

Comment:
aha, so you do not compute directly forces by FFT? Perhaps it is fine. I actually don't know why I did all the hassle to compute forces by FFT, but as I remember I had some problem with numberical noise from finite-differences which was suppressed by using FFT direcly

Reply:
In my testing there is a little bit of difference between the FFT derivative and the finite difference derivative, but the final force field is almost identical.  I would maybe like to test at some point if using a 5-point stencil instead of 3 makes a difference.